src/main/java/telraam/logic/robustLapper/robustLapperIncremental/RobustLapperIncremental.java

```diff
@@ -0,0 +1,226 @@
+import telraam.database.models.*;
+import telraam.logic.Lapper;
+
+import java.sql.Timestamp;
+import java.util.*;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+// 3 Potential problems:
+// 1: Telraam gets restarted
+//          -> lastDetection & lastSwitchoverId are set back to 0 and all previous laps get deleted
+// 2: A switchover gets added with a timestamp earlier than the last detection that was handled ergo some detections ended up in the wrong team
+//          -> processData checks for switchover timestamp. If it's earlier it fetches all detections since second last registered lap before the switchover timestamp
+//          -> team class gets a detection with timestamp earlier than its latest detection, deletes all laps since then and fetches the detections again
+// 3: Some detections don't get added in chronological order (e.g. temporarily ronny disconnect)
+//          -> See second point for 2:
+
+// Implement Lapper for easier use in App and Fetcher
+public class RobustLapperIncremental implements Lapper {
+
+    private final Jdbi jdbi;
+
+    // Data processing variables
+    private final int MIN_RSSI = -85;                                   // Min RSSI
+    private Detection lastDetection = new Detection(-1, new Timestamp(0));  // Last detection
+    private int lastSwitchoverId = -1;                                  // ID of last switchover
+    private Map<Integer, Team> teamById;                                // Map teamId -> team
+    private Map<Integer, RobustLapperTeam> teamLappers;                 // Map teamId -> RobustLapperTeam class
+    private Map<Integer, List<Detection>> teamDetections;               // Map teamId -> their detections
+
+    // Logger
+    private final Logger logger;
+
+    // Lap source variables
+    private final String SOURCE_NAME = "robust-lapper-incremental";     // Name of lap source
+    private int lapSourceId;                                            // ID in database
+
+    // Scheduler variables
+    private final int DEBOUNCE_TIMEOUT = 10;
+    private final ScheduledExecutorService scheduler;
+    private boolean debounceScheduled;
+
+    public RobustLapperIncremental(Jdbi jdbi) {
+        this.jdbi = jdbi;
+
+        this.logger = Logger.getLogger(RobustLapperIncremental.class.getName());
+
+        this.scheduler = Executors.newScheduledThreadPool(1);
+        this.debounceScheduled = false;
+
+        // Get the lapSourceId, create the source if needed
+        LapSourceDAO lapSourceDAO = jdbi.onDemand(LapSourceDAO.class);
+        lapSourceDAO.getByName(SOURCE_NAME).ifPresentOrElse(
+                lapSource -> this.lapSourceId = lapSource.getId(),
+                () -> this.lapSourceId = lapSourceDAO.insert(new LapSource(SOURCE_NAME))
+        );
+
+        setTeamVariables();
+        deleteLaps();
```

Comment:
If Telraam is restarted, I think all laps will be deleted, which is not desired

Reply:
Now, when telraam gets restarted, all laps entered in the database by RobustLapperIncremental are deleted. The last detection is set "0", same for the last switchover. This means that with the next new detection it'll fetch everything from the database but after that it'll only fetch the new detections again. While this is not ideal, I still think it's fine as it only fetches everything once.

If you immediately want to work with only the new detections you’d have to delete the last two laps from each team. Look which team’s lap was set the earliest in timestamp and then get the detection that corresponds the closest in timestamp with that lap. Or you can just delete every lap entered and fetch all detections just once ;)